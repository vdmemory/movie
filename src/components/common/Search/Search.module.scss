.bar {
    position: relative;
    margin: 0 auto;
    max-width: 575px;
    border-radius: 30px;
    border: 1px solid #dcdcdc;
    background: white;

    &:hover {
        box-shadow: 1px 1px 8px 1px #dcdcdc;
    }

    &:focus-within {
        box-shadow: 1px 1px 8px 1px #dcdcdc;
        outline: none;
    }

    & .searchbar {
        height: 45px;
        border: none;
        width: calc(100% - 70px);
        font-size: 16px;
        outline: none;
        margin-left: 50px;
        border-radius: 30px;
    }

    & .searchIcon {
        width: 20px;
        height: 20px;
        position: absolute;
        top: 13px;
        left: 18px;
        color: #606060;
    }
}