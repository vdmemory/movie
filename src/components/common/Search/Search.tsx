import React from "react";
import { BsSearch } from "react-icons/bs";
import debounce from "debounce";

import styles from "./Search.module.scss";
import { useDispatch } from "react-redux";
import { getSearchMoviesRequest } from "../../../store/actions";

type Props = {
    placeholder: string;
    title: string;
    type: string
};

const Search: React.FC<Props> = ({ placeholder, title, type }) => {
    const dispatch = useDispatch();

    const onChangeDispatch = (value: string) => {
        dispatch(getSearchMoviesRequest(value));
    };

    const debounceChange = debounce(onChangeDispatch, 500);

    const onSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        debounceChange(e.target.value);
    };

    return (
        <div className={styles.bar}>
            <BsSearch className={styles.searchIcon} />
            <input
                className={styles.searchbar}
                placeholder={placeholder}
                type={type}
                title={title}
                onChange={onSearchChange}
            />
        </div>
    );
};

export default Search;
