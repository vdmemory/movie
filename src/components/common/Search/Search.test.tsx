import React, { Dispatch } from "react";
import * as redux from "react-redux";
import { createRenderer } from "react-test-renderer/shallow";
import { debounce } from "debounce";

import Search from "./Search";
import { BsSearch } from "react-icons/bs";

type Props = {
    placeholder: string;
    title: string;
    type: string;
    dispatch: Dispatch<any>;
};

const spySelector = jest.spyOn(redux, "useSelector");
spySelector.mockReturnValue({ searchValue: "test" });

const mockDebounce = jest.spyOn(debounce, "debounce");
mockDebounce.mockImplementation((fn: any) => fn);

const setup = (props: Props) => {
    const renderer = createRenderer();
    renderer.render(
        <Search {...props} />
    );
    const output = renderer.getRenderOutput();
    return { output, props, renderer };
};

const propsRender: Props = {
    placeholder: "this placeholder",
    title: "this title",
    type: "this type",
    dispatch: jest.fn()
};

describe("Search Component", () => {
    describe("Render component and its components", () => {
        test("Should renders correctly", () => {
            const { output } = setup(propsRender);
            expect(output).toMatchSnapshot();
        });

        test("Should render container and has a class", () => {
            const { output } = setup(propsRender);
            expect(output.type).toBe("div");
            expect(output.props.className).toBe("bar");
        });

        test("Should render input and has a class", () => {
            const { output } = setup(propsRender);
            const [, input] = output.props.children;
            expect(input.type).toBe("input");
            expect(input.props.className).toBe("searchbar");
        });

        test("Should call handleSearchChange on change", () => {
            const event = {
                preventDefault() {},
                target: { value: "test" },
            };

            const { output, props } = setup(propsRender);
            const [, input] = output.props.children;
            input.props.onChange(event);

            expect(props.dispatch).toBeCalled();
            expect(spySelector).toBeCalled();
            expect(mockDebounce).toBeCalled();
        });

        describe("BsSearch", () => {
            test("Should render and check class", () => {
                const { output } = setup(propsRender);
                expect(output.props.children[0].type).toBe(BsSearch);
                expect(output.props.children[0].props.className).toBe(
                    "searchIcon"
                );
            });
        });
    });
});
