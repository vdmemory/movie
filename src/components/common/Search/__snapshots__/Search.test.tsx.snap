// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component Render component and its components Should renders correctly 1`] = `
<div
  className="bar"
>
  <BsSearch
    className="searchIcon"
  />
  <input
    className="searchbar"
    onChange={[Function]}
    placeholder="this placeholder"
    title="this title"
    type="this type"
    value="test"
  />
</div>
`;
